File: manual_examples/FilterAndFoldInts.hs
module FilterAndFoldInts where

import Prelude hiding (foldl, (.), filter)

foldl :: (Int -> Int -> Int) -> Int -> [Int] -> Int
foldl _ b []     = b
foldl f b (x:xs) = x `f` foldl f b xs

(.) :: ([Int] -> Int) -> ([Int] -> [Int]) -> [Int] -> Int
(.) f g = \x -> f (g x)

filter :: (Int -> Bool) -> [Int] -> [Int]
filter _ [] = []
filter pred (x:xs) = if pred x then x : (filter pred xs) else filter pred xs
  -- | pred x = x : filter pred xs
  -- | otherwise = filter pred xs
          
-- Testing for point-free style in this example
filterAndFoldInts :: (Int -> Bool) -> (Int -> Int -> Int) -> Int -> [Int] -> Int
filterAndFoldInts filt f base x = ((foldl f base) . (filter filt)) x


                                

File: manual_examples/FilterAndFoldInts.pl
:- module(_,_,[functional, hiord]).
:- use_module('~/hs-to-ciao/lib/ciao_prelude.pl').

:- meta_predicate sum(?,?,?).
sum(X,Y) := +(X,Y).

:- meta_predicate greater(?,?,?).
greater(X,Y) := ~bool(X>Y).

:- meta_predicate filterAndFoldInts(pred(2),pred(3),?,?,?).
filterAndFoldInts(Filt, F, Base, List) := ~compose(foldl(F, Base), filter(Filt), List).
